File: lib/controllers/device_vital_controller.dart
import 'dart:convert';

import 'package:docuhealth/model/device_vital/device_vital.dart';
import 'package:docuhealth/services/base_client.dart';
import 'package:flutter/material.dart';
import 'package:intl/intl.dart';

/// Controller for device vitals that come from BLE hardware via MQTT
class DeviceVitalController extends ChangeNotifier {
  final BaseClient _baseClient = BaseClient();

  // Latest vitals for dashboard display
  LatestVitalsResponse? latestVitals;

  // Historical data
  List<DeviceVital> vitalHistory = [];

  // Analytics data
  VitalAnalyticsData? analyticsData;

  // Linked devices
  List<DeviceInfo> devices = [];

  // Loading states
  bool isLoadingLatest = false;
  bool isLoadingHistory = false;
  bool isLoadingAnalytics = false;
  bool isLoadingDevices = false;

  // Error messages
  String? errorMessage;

  /// Fetch latest vitals for all types (dashboard view)
  Future<void> fetchLatestVitals({String? deviceName}) async {
    isLoadingLatest = true;
    errorMessage = null;
    notifyListeners();

    try {
      String endpoint = 'device-vitals/latest';
      if (deviceName != null && deviceName.isNotEmpty) {
        endpoint += '?device_name=$deviceName';
      }

      final response = await _baseClient.get(endpoint, true);
      final jsonResponse = jsonEncode(response);
      latestVitals = getLatestVitalsFromJson(jsonResponse);

      if (!latestVitals!.success) {
        errorMessage = latestVitals!.message;
      }
    } catch (e) {
      errorMessage = 'Failed to fetch latest vitals: $e';
    }

    isLoadingLatest = false;
    notifyListeners();
  }

  /// Fetch vital history with optional filters
  Future<void> fetchVitalHistory({
    String? vitalType,
    String? deviceName,
    DateTime? date,
    DateTime? startDate,
    DateTime? endDate,
    int limit = 100,
  }) async {
    isLoadingHistory = true;
    errorMessage = null;
    notifyListeners();

    try {
      final params = <String>[];
      
      if (vitalType != null) {
        params.add('vital_type=$vitalType');
      }
      if (deviceName != null) {
        params.add('device_name=$deviceName');
      }
      if (date != null) {
        params.add('date=${DateFormat('yyyy-MM-dd').format(date)}');
      }
      if (startDate != null) {
        params.add('start_date=${DateFormat('yyyy-MM-dd').format(startDate)}');
      }
      if (endDate != null) {
        params.add('end_date=${DateFormat('yyyy-MM-dd').format(endDate)}');
      }
      params.add('limit=$limit');

      final queryString = params.isNotEmpty ? '?${params.join('&')}' : '';
      final response = await _baseClient.get('device-vitals$queryString', true);
      final jsonResponse = jsonEncode(response);
      final result = getDeviceVitalsFromJson(jsonResponse);

      if (result.success) {
        vitalHistory = result.data;
      } else {
        errorMessage = result.message;
        vitalHistory = [];
      }
    } catch (e) {
      errorMessage = 'Failed to fetch vital history: $e';
      vitalHistory = [];
    }

    isLoadingHistory = false;
    notifyListeners();
  }

  /// Fetch analytics for a specific vital type
  Future<void> fetchAnalytics({
    required String vitalType,
    String period = 'day',
    DateTime? startDate,
    DateTime? endDate,
  }) async {
    isLoadingAnalytics = true;
    errorMessage = null;
    notifyListeners();

    try {
      final params = <String>['vital_type=$vitalType', 'period=$period'];
      
      if (startDate != null) {
        params.add('start_date=${DateFormat('yyyy-MM-dd').format(startDate)}');
      }
      if (endDate != null) {
        params.add('end_date=${DateFormat('yyyy-MM-dd').format(endDate)}');
      }

      final queryString = '?${params.join('&')}';
      final response = await _baseClient.get('device-vitals/analytics$queryString', true);
      final jsonResponse = jsonEncode(response);
      final result = getVitalAnalyticsFromJson(jsonResponse);

      if (result.success) {
        analyticsData = result.data;
      } else {
        errorMessage = result.message;
        analyticsData = null;
      }
    } catch (e) {
      errorMessage = 'Failed to fetch analytics: $e';
      analyticsData = null;
    }

    isLoadingAnalytics = false;
    notifyListeners();
  }

  /// Fetch list of linked devices
  Future<void> fetchDevices() async {
    isLoadingDevices = true;
    errorMessage = null;
    notifyListeners();

    try {
      final response = await _baseClient.get('device-vitals/devices', true);
      final jsonResponse = jsonEncode(response);
      final result = getDevicesFromJson(jsonResponse);

      if (result.success) {
        devices = result.data;
      } else {
        errorMessage = result.message;
        devices = [];
      }
    } catch (e) {
      errorMessage = 'Failed to fetch devices: $e';
      devices = [];
    }

    isLoadingDevices = false;
    notifyListeners();
  }

  /// Link a device to the current profile
  Future<bool> linkDevice(String deviceName) async {
    try {
      final response = await _baseClient.post(
        'device-vitals/link-device',
        {'device_name': deviceName},
        true,
      );

      if (response['success'] == true) {
        await fetchDevices();
        return true;
      } else {
        errorMessage = response['message'];
        return false;
      }
    } catch (e) {
      errorMessage = 'Failed to link device: $e';
      return false;
    }
  }

  /// Store a vital reading directly (from the POC app)
  Future<bool> storeVital({
    required String deviceName,
    required String vitalType,
    required double value,
    DateTime? recordedAt,
  }) async {
    try {
      final payload = {
        'device_name': deviceName,
        'vital_type': vitalType,
        'value': value,
        if (recordedAt != null) 'recorded_at': recordedAt.toIso8601String(),
      };

      final response = await _baseClient.post('device-vitals', payload, true);

      if (response['success'] == true) {
        // Refresh latest vitals after storing
        await fetchLatestVitals();
        return true;
      } else {
        errorMessage = response['message'];
        return false;
      }
    } catch (e) {
      errorMessage = 'Failed to store vital: $e';
      return false;
    }
  }

  /// Store vital in MQTT format (from the POC app gateway)
  Future<bool> storeVitalMqttFormat({
    required String deviceName,
    required String vitalType,
    required dynamic value,
  }) async {
    try {
      // Build MQTT format payload: {device_name: [{vital_type: value}]}
      final payload = {
        deviceName: [
          {vitalType: value}
        ]
      };

      final response = await _baseClient.post('device-vitals', payload, true);

      if (response['success'] == true) {
        await fetchLatestVitals();
        return true;
      } else {
        errorMessage = response['message'];
        return false;
      }
    } catch (e) {
      errorMessage = 'Failed to store vital: $e';
      return false;
    }
  }

  /// Delete a vital reading
  Future<bool> deleteVital(int id) async {
    try {
      final response = await _baseClient.get('device-vitals/delete/$id', true);

      if (response['success'] == true) {
        vitalHistory.removeWhere((v) => v.id == id);
        notifyListeners();
        return true;
      } else {
        errorMessage = response['message'];
        return false;
      }
    } catch (e) {
      errorMessage = 'Failed to delete vital: $e';
      return false;
    }
  }

  /// Get formatted display string for a vital type's latest value
  String getLatestValueDisplay(String vitalType) {
    if (latestVitals == null || latestVitals!.data[vitalType] == null) {
      return '--';
    }

    final vital = latestVitals!.data[vitalType]!;
    final unit = vital.unit ?? '';

    if (vitalType == DeviceVital.typeBodyTemp) {
      return '${vital.value.toStringAsFixed(2)} $unit';
    }
    return '${vital.value.toStringAsFixed(0)} $unit';
  }

  /// Check if any vitals are available
  bool get hasAnyVitals {
    if (latestVitals == null) return false;
    return latestVitals!.data.values.any((v) => v != null);
  }

  /// Clear all data
  void clear() {
    latestVitals = null;
    vitalHistory = [];
    analyticsData = null;
    devices = [];
    errorMessage = null;
    notifyListeners();
  }
}

File: Backend/app/Http/Controllers/API/DeviceVitalController.php
<?php

namespace App\Http\Controllers\API;

use App\Http\Controllers\Controller;
use Illuminate\Http\Request;
use Illuminate\Support\Facades\Validator;
use App\Library\Structure;
use App\Models\DeviceVital;
use Carbon\Carbon;
use DB;

class DeviceVitalController extends Controller
{
    use Structure;

    /**
     * Store a vital reading from the device/app
     * 
     * Expected payload format (matching MQTT):
     * {
     *   "device_name": "DeviceName",
     *   "vital_type": "SpO2",
     *   "value": 97
     * }
     * 
     * OR MQTT format:
     * {
     *   "DeviceName": [{"SpO2": 97}]
     * }
     */
    public function store(Request $request)
    {
        $profile_id = $request->profile->id ?? null;

        // Check if it's MQTT format (device_name as key)
        $rawData = $request->all();
        $firstKey = array_key_first($rawData);
        
        if ($firstKey && is_array($rawData[$firstKey] ?? null)) {
            // MQTT format: {device_name: [{vital_type: value}]}
            $payload = json_encode($rawData);
            $created = DeviceVital::createFromMqttPayload($payload, $profile_id);
            
            if (empty($created)) {
                return response()->json($this->structure(false, 'Invalid payload format'), 200);
            }
            
            return response()->json($this->structure(true, 'Vital(s) saved', [
                'count' => count($created),
                'vitals' => $created
            ]), 200);
        }

        // Standard format
        $validator = Validator::make($request->all(), [
            'device_name' => 'required|string|max:100',
            'vital_type' => 'required|string|in:' . implode(',', DeviceVital::validTypes()),
            'value' => 'required|numeric',
            'recorded_at' => 'nullable|date',
        ]);

        if ($validator->fails()) {
            return response()->json($this->structure(false, $validator->errors()->first()), 200);
        }

        try {
            $value = $request->value;
            
            // Handle temperature (may be sent as value * 10)
            if ($request->vital_type === DeviceVital::TYPE_BODY_TEMP && $value > 100) {
                $value = $value / 10.0;
            }

            $vital = DeviceVital::create([
                'profile_id' => $profile_id,
                'device_name' => $request->device_name,
                'vital_type' => $request->vital_type,
                'value' => $value,
                'unit' => DeviceVital::UNITS[$request->vital_type] ?? null,
                'recorded_at' => $request->recorded_at ? Carbon::parse($request->recorded_at) : now(),
            ]);

            return response()->json($this->structure(true, 'Vital saved', $vital), 200);

        } catch (\Exception $e) {
            return response()->json($this->structure(false, 'Failed to save vital: ' . $e->getMessage()), 200);
        }
    }

    /**
     * Store vitals via MQTT webhook (public endpoint for MQTT broker callback)
     * No authentication required - data comes directly from MQTT broker
     */
    public function mqttWebhook(Request $request)
    {
        $payload = $request->getContent();
        
        // Log incoming webhook
        \Log::info('MQTT Webhook received', ['payload' => $payload]);
        
        $created = DeviceVital::createFromMqttPayload($payload);
        
        if (empty($created)) {
            return response()->json(['success' => false, 'message' => 'No valid vitals in payload'], 400);
        }

        return response()->json([
            'success' => true,
            'message' => 'Vitals saved',
            'count' => count($created)
        ], 200);
    }

    /**
     * Get vitals for the authenticated profile
     */
    public function get(Request $request)
    {
        $profile_id = $request->profile->id ?? null;
        
        if (!$profile_id) {
            return response()->json($this->structure(false, 'Profile not found'), 200);
        }

        $query = DeviceVital::profile($profile_id);

        // Filter by device name
        if ($request->device_name) {
            $query->device($request->device_name);
        }

        // Filter by vital type
        if ($request->vital_type) {
            $query->type($request->vital_type);
        }

        // Filter by date
        if ($request->date) {
            $query->whereDate('recorded_at', Carbon::parse($request->date));
        }

        // Filter by date range
        if ($request->start_date) {
            $query->dateRange(
                Carbon::parse($request->start_date),
                $request->end_date ? Carbon::parse($request->end_date) : null
            );
        }

        // Order and limit
        $query->orderBy('recorded_at', 'desc');
        
        $limit = min($request->limit ?? 100, 500);
        $vitals = $query->limit($limit)->get();

        return response()->json($this->structure(true, 'Device vitals', $vitals), 200);
    }

    /**
     * Get latest vitals for all types (dashboard view)
     */
    public function latest(Request $request)
    {
        $profile_id = $request->profile->id ?? null;
        
        if (!$profile_id) {
            return response()->json($this->structure(false, 'Profile not found'), 200);
        }

        $deviceName = $request->device_name;
        
        // Get latest reading for each vital type
        $latestVitals = DeviceVital::profile($profile_id)
            ->when($deviceName, fn($q) => $q->device($deviceName))
            ->select('vital_type', DB::raw('MAX(id) as latest_id'))
            ->groupBy('vital_type')
            ->pluck('latest_id');

        $vitals = DeviceVital::whereIn('id', $latestVitals)
            ->orderBy('vital_type')
            ->get()
            ->keyBy('vital_type');

        // Build response with all vital types
        $result = [];
        foreach (DeviceVital::validTypes() as $type) {
            $vital = $vitals->get($type);
            $result[$type] = $vital ? [
                'value' => $vital->value,
                'unit' => $vital->unit,
                'device_name' => $vital->device_name,
                'recorded_at' => $vital->recorded_at,
            ] : null;
        }

        return response()->json($this->structure(true, 'Latest vitals', $result), 200);
    }

    /**
     * Get analytics/history for a specific vital type
     */
    public function analytics(Request $request)
    {
        $profile_id = $request->profile->id ?? null;
        
        if (!$profile_id) {
            return response()->json($this->structure(false, 'Profile not found'), 200);
        }

        $validator = Validator::make($request->all(), [
            'vital_type' => 'required|in:' . implode(',', DeviceVital::validTypes()),
            'period' => 'nullable|in:day,week,month,custom',
            'start_date' => 'nullable|date',
            'end_date' => 'nullable|date',
        ]);

        if ($validator->fails()) {
            return response()->json($this->structure(false, $validator->errors()->first()), 200);
        }

        $vitalType = $request->vital_type;
        $period = $request->period ?? 'day';
        
        // Calculate date range based on period
        $endDate = Carbon::today();
        switch ($period) {
            case 'week':
                $startDate = Carbon::today()->subDays(6);
                break;
            case 'month':
                $startDate = Carbon::today()->subDays(29);
                break;
            case 'custom':
                $startDate = $request->start_date ? Carbon::parse($request->start_date) : Carbon::today();
                $endDate = $request->end_date ? Carbon::parse($request->end_date) : Carbon::today();
                break;
            default: // day
                $startDate = Carbon::today();
                break;
        }

        $vitals = DeviceVital::profile($profile_id)
            ->type($vitalType)
            ->whereDate('recorded_at', '>=', $startDate)
            ->whereDate('recorded_at', '<=', $endDate)
            ->orderBy('recorded_at', 'asc')
            ->get();

        // Calculate statistics
        $values = $vitals->pluck('value');
        $stats = [
            'min' => $values->min(),
            'max' => $values->max(),
            'avg' => round($values->avg(), 2),
            'count' => $vitals->count(),
        ];

        // Group by date for chart data
        $grouped = $vitals->groupBy(fn($v) => Carbon::parse($v->recorded_at)->format('Y-m-d'));
        $chartData = [];
        foreach ($grouped as $date => $dateVitals) {
            $chartData[] = [
                'date' => $date,
                'readings' => $dateVitals->map(fn($v) => [
                    'time' => Carbon::parse($v->recorded_at)->format('H:i'),
                    'value' => $v->value,
                ])->values(),
            ];
        }

        return response()->json($this->structure(true, 'Vital analytics', [
            'vital_type' => $vitalType,
            'unit' => DeviceVital::UNITS[$vitalType] ?? null,
            'period' => $period,
            'start_date' => $startDate->format('Y-m-d'),
            'end_date' => $endDate->format('Y-m-d'),
            'statistics' => $stats,
            'data' => $chartData,
        ]), 200);
    }

    /**
     * Delete a vital reading
     */
    public function delete(Request $request, $id)
    {
        $profile_id = $request->profile->id ?? null;
        
        if (!$profile_id) {
            return response()->json($this->structure(false, 'Profile not found'), 200);
        }

        $vital = DeviceVital::profile($profile_id)->find($id);
        
        if (!$vital) {
            return response()->json($this->structure(false, 'Vital not found'), 200);
        }

        $vital->delete();

        return response()->json($this->structure(true, 'Vital deleted'), 200);
    }

    /**
     * Link a device to the current profile
     * This allows associating existing device readings with a profile
     */
    public function linkDevice(Request $request)
    {
        $profile_id = $request->profile->id ?? null;
        
        if (!$profile_id) {
            return response()->json($this->structure(false, 'Profile not found'), 200);
        }

        $validator = Validator::make($request->all(), [
            'device_name' => 'required|string|max:100',
        ]);

        if ($validator->fails()) {
            return response()->json($this->structure(false, $validator->errors()->first()), 200);
        }

        // Update all unlinked vitals from this device to the current profile
        $updated = DeviceVital::where('device_name', $request->device_name)
            ->whereNull('profile_id')
            ->update(['profile_id' => $profile_id]);

        return response()->json($this->structure(true, 'Device linked', [
            'device_name' => $request->device_name,
            'linked_readings' => $updated,
        ]), 200);
    }

    /**
     * Get list of known devices for the profile
     */
    public function devices(Request $request)
    {
        $profile_id = $request->profile->id ?? null;
        
        if (!$profile_id) {
            return response()->json($this->structure(false, 'Profile not found'), 200);
        }

        $devices = DeviceVital::profile($profile_id)
            ->select('device_name', DB::raw('COUNT(*) as reading_count'), DB::raw('MAX(recorded_at) as last_reading'))
            ->groupBy('device_name')
            ->orderBy('last_reading', 'desc')
            ->get();

        return response()->json($this->structure(true, 'Devices', $devices), 200);
    }
}
